app/src/main/java/com/p2p/presentation/home/HomeActivity.kt

```diff
@@ -1,25 +1,65 @@
 package com.p2p.presentation.home
 
+import android.Manifest.permission.ACCESS_COARSE_LOCATION
+import android.Manifest.permission.ACCESS_FINE_LOCATION
+import android.content.pm.PackageManager.PERMISSION_GRANTED
 import android.os.Bundle
 import android.view.WindowManager
+import androidx.core.app.ActivityCompat
+import androidx.core.content.ContextCompat
+import com.google.android.material.dialog.MaterialAlertDialogBuilder
 import com.p2p.R
 import com.p2p.presentation.base.BaseActivity
 import com.p2p.presentation.home.games.GamesFragment
 
-// TODO: should we ask ACCESS_COARSE_LOCATION permission on runtime? on some examples they do, maybe it's the reason
-//  of why the discovering isn't working, but we didn't investigate it so much.
 class HomeActivity : BaseActivity() {
 
+    private var hasShownLocationExplanation = false
+
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         if (savedInstanceState == null) {
             addFragment(GamesFragment.newInstance(), shouldAddToBackStack = false)
             removeSplashStyle()
         }
+        if (!hasLocationPermissions()) requestLocationPermissions()
+    }
+
+    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        if (requestCode == REQUEST_LOCATION_PERMISSION_CODE && grantResults.any { it != PERMISSION_GRANTED }) {
+            if (!hasShownLocationExplanation) {
+                hasShownLocationExplanation = true
+                MaterialAlertDialogBuilder(this)
+                    .setTitle(R.string.request_location_title)
+                    .setMessage(R.string.request_location_description)
+                    .setPositiveButton(R.string.understood) { _, _ -> requestLocationPermissions() }
+                    .show()
+            } else {
```

Comment:
este else no sería el del ` if (requestCode == REQUEST_LOCATION_PERMISSION_CODE && grantResults.any { it != PERMISSION_GRANTED })` ? quedó como else del `!hasShownLocationExplanation`

Reply:
Tengo entendido que estaría bien así, porque cuando ya se dió la explicación del por qué se requiere, y le usuarie no le dio el permiso requerido, recién ahi se cierra la aplicación.

Si este else quedara en el primer if, entonces no se llegaría a mostrar la explicación

Reply:
Claro, lo que dice Vale. El primer if chequea si es el pedido de permisos de localización y si no lo aceptaron. En caso de que eso se cumpla, la primera vez mostramos un cartel y la segunda cerramos la app.

Reply:
Claro, lo que dice Vale. El primer if chequea si es el pedido de permisos de localización y si no lo aceptaron. En caso de que eso se cumpla, la primera vez mostramos un cartel y la segunda cerramos la app.